src/part-guide/adv-async-await.md

```diff
@@ -2,26 +2,142 @@
+
+
 ## Blocking and cancellation
 
-- Two important concepts to be aware of early, we'll revisit in more detail as we go along
-- Cancellation
-  - How to do it
-    - drop a future
-    - cancellation token
-    - abort functions
-  - Why it matters, cancellation safety (forward ref)
-- Blocking
-  - IO and computation can block
-  - why it's bad
-  - how to deal is a forward ref to io chapter
+Blocking and cancellation are important to keep in mind when programming with async Rust. These concepts are not localised to any particular feature or function, but are ubiquitous properties of the system which you must understand to write correct code.
+
+### Blocking IO
+
+We say a thread (note we're talking about OS threads here, not async tasks) is blocked when it can't make any progress. That's usually because it is waiting for the OS to complete a task on it's behalf (usually IO). Importantly, while a thread is blocked, the OS knows not to schedule it so that other threads can make progress. This is fine in a multithreaded program because it lets other threads make progress while the blocked thread is waiting. However, in an async program, there are other tasks which should be scheduled on the same OS thread, but the OS doesn't know about those and keeps the whole thread waiting. This means that rather than the single task waiting for it's IO to complete (which is fine), many tasks have to wait (not fine).
+
+We'll talk soon about non-blocking/async IO. For now, just know that non-blocking IO is IO which the async runtime knows about and so will only block the task which is waiting for it, not the whole thread. It is very important to only use non-blocking IO from an async task, never blocking IO (which is the only kind provided in Rust's standard library).
+
+### Blocking computation
+
+You can also block the thread by doing computation (this is not quite the same as blocking IO, since the OS is not involved, but the effect is similar). If you have long-running computation (with or without blocking IO) without yielding control to the runtime, then that task will never give the scheduler a chance to schedule other tasks. Remember that async programming uses cooperative multitasking? Here a task is not cooperating, so other tasks won't get a chance to get work done. We'll discuss ways to mitigate this later.
+
+There are many other ways to block a whole thread, and we'll come back to blocking several times in this guide.
+
+### Cancellation
+
+Cancellation means stopping a future (or task) from executing. Since in Rust, futures must be driven forward by an external force (like the async runtime), if a future is no longer driven forward then it will not execute any more. If a future is dropped (remember, a future is just a plain old Rust object), then it can never make any more progress and is cancelled.
```

Comment:
Makes sense in general, in this case though I think the first 'in Rust' is warranted because I want this is a behaviour which is pretty unique to Rust and in contrast to other systems (and this is a known footgun for those coming from other languages).

The second use is a bit of an in-joke referencing POJOs: https://en.wikipedia.org/wiki/Plain_old_Java_object

Reply:
> Makes sense in general, in this case though I think the first 'in Rust' is warranted because I want this is a behaviour which is pretty unique to Rust and in contrast to other systems (and this is a known footgun for those coming from other languages).

Yes... when making the suggestion I'd almost gone the other way for that reason and suggested spelling that out instead.  Maybe that is the thing to do, is to add a sentence or two saying that outright.

> The second use is a bit of an in-joke referencing POJOs: https://en.wikipedia.org/wiki/Plain_old_Java_object

Ah, yes.  Well, carry on then.